README.md
# PyGCN

![Python Package Index status](https://img.shields.io/pypi/v/pygcn)
[![Code coverage status](https://codecov.io/gh/lpsinger/pygcn/branch/main/graph/badge.svg?token=QDu6W3LiF6)](https://codecov.io/gh/lpsinger/pygcn)

Anonymous VOEvent client for receiving GCN/TAN notices in XML format

The [Gamma-ray Coordinates Network/Transient Astronomy Network (GCN/TAN)][1] is
a system for distributing astronomical alerts, largely focused on operations of
and detections from high-energy satellite missions.

GCN/TAN disseminates both Notices (prompt, machine-readable alerts) and
Circulars (human-readable correspondence) through a handful of delivery methods
and formats.

This package implements a simple client that listens for [VOEvent][2] XML
format notices over the custom TCP/IP [VOEvent Transport Protocol][3]. By
default, it connects to one of the anonymous GCN/TAN server, so no sign-up
or configuration is necessary to begin receiving alerts.

## Installation

To install PyGCN, simply run:

    $ pip install --user pygcn

## Usage

PyGCN provides an example script called `pygcn-listen` that will simply write
all VOEvents that it receives to files in the current directory. To try it out,
simply run:

    $ pygcn-listen

and then type Control-C to quit.

## Writing a custom GCN handler

You can also write your own handler that performs a custom action for every GCN
that is received. A handler function takes two arguments: `payload`, the raw
content of the GCN, and `root`, the root element of the XML document as parsed
by [`lxml.etree`][5]. Here is a basic example:

```python
#!/usr/bin/env python
import gcn

# Define your custom handler here.
def handler(payload, root):
    # Get the IVORN, or unique VOEvent ID, and print it.
    print(root.attrib['ivorn'])

    # Print all of the event attributes.
    for param in root.findall('./What/Param'):
        name = param.attrib['name']
        value = param.attrib['value']
        print('{} = {}'.format(name, value))

# Listen for VOEvents until killed with Control-C.
gcn.listen(handler=handler)
```
## Filtering

You can also filter events by notice type using
`gcn.include_notice_types` or `gcn.exclude_notice_types`.
Here is an example:

```python
#!/usr/bin/env python
import gcn

# Define your custom handler here.
@gcn.include_notice_types(
    gcn.notice_types.FERMI_GBM_FLT_POS,  # Fermi GBM localization (flight)
    gcn.notice_types.FERMI_GBM_GND_POS,  # Fermi GBM localization (ground)
    gcn.notice_types.FERMI_GBM_FIN_POS)  # Fermi GBM localization (final)
def handler(payload, root):
    # Look up right ascension, declination, and error radius fields.
    pos2d = root.find('.//{*}Position2D')
    ra = float(pos2d.find('.//{*}C1').text)
    dec = float(pos2d.find('.//{*}C2').text)
    radius = float(pos2d.find('.//{*}Error2Radius').text)

    # Print.
    print('ra = {:g}, dec={:g}, radius={:g}'.format(ra, dec, radius))

# Listen for VOEvents until killed with Control-C.
gcn.listen(handler=handler)
```


[1]: http://gcn.gsfc.nasa.gov
[2]: http://www.ivoa.net/documents/VOEvent
[3]: http://www.ivoa.net/documents/Notes/VOEventTransport
[4]: https://docs.python.org/2/library/xml.etree.elementtree.html
[5]: http://lxml.de


ChangeLog
Changelog
=========

1.1.3 (unreleased)
------------------

- The ``@include_notice_type`` and ``@exclude_notice_type`` decorators now pass
  through all arguments to the wrapped function. [David Palmer]

1.1.2 (2021-12-22)
------------------

- Run unit tests under Python 3.10. [Leo Singer]

- Drop support for Python 3.6, which reached end of life on 2021-12-23.
  [Leo Singer]

- Remove setup.py file. [Leo Singer]

1.1.1 (2021-04-20)
------------------

- Remove use of the third-party ``pkg_resources`` module from unit tests and
  use Python standard library module ``importlib.resources`` instead.
  [Leo Singer]

- Add new notice type: ``SK_SN``. [Leo Singer]

- Rename the development branch from ``master`` to ``main``. [Leo Singer]

1.1.0 (2021-01-29)
------------------

- Add new notice types: ``AMON_NU_EM_COINC``, ``ICECUBE_CASCADE``.
 [Volodymyr Savchenko]

- Drop support for Python 2.7 and 3.5, which have reached end-of-life.
  Add support for Python 3.9. [Leo Singer]

1.0.2 (2020-03-05)
------------------

- Fix incorrect license filename in setup.cfg. [Leo Singer]

1.0.1 (2020-03-05)
------------------

- Add new notice type: ``LVC_EARLY_WARNING``. [Leo Singer]

- Update license to GPL-3+. [Leo Singer]

1.0.0 (2020-01-29)
------------------

- The ``gcn.listen`` method can now cycle through a list of remote hosts when
  re-establishing the connection to GCN. [Varun Bhalerao, Aaron Jones]

- Add new notice types: ``AGILE_MCAL_ALERT``, ``GWHEN_COINC``. [Leo Singer]

0.1.20 (2019-02-12)
-------------------

- Add new notice types: ``HAWC_BURST_MONITOR``, ``ICECUBE_ASTROTRACK_GOLD``,
  and ``ICECUBE_ASTROTRACK_BRONZE``. [Vladimir Savchenko]

0.1.19 (2019-02-12)
-------------------

- Add the ``LVC_RETRACTION`` notice type. [Leo Singer]

0.1.18 (2018-12-11)
-------------------

- Add ``{http://telescope-networks.org/xml/Transport/v1.1}Transport`` and
  ``{http://telescope-networks.org/schema/Transport/v1.1}Transport`` as
  recognized VOEvent Transport Protocol namespaces to support the 4pisky and
  comet brokers. [reported by Martin Dyer]

0.1.17 (2018-10-04)
-------------------

- Add missing dependency on six. [Leo Singer]

- Fix handling of payload command line arguments for pygcn-server.
  Thanks go out to WeiKang Zheng of UC Berkeley for reporting this.
  [Leo Singer]

0.1.16 (2018-03-03)
-------------------

- Add ``functools.wraps()`` to handler decorators so that the original
  function docstrings are preserved. [Leo Singer]

0.1.15 (2018-02-10)
-------------------

- Update module __all__ variables for convenience. [Leo Singer]

- Add experimental PEP 435 style enum class, ``gcn.NoticeType``. [Leo Singer]

0.1.14 (2018-02-01)
-------------------

- Depend unconditionally on lxml. Some of the examples depend on it because
  of differences in how xml.etree and lxml.etree handle XML namespaces.
  [Leo Singer]

0.1.13 (2018-01-24)
-------------------

- Include missing test data in package. [Leo Singer]

0.1.12 (2018-01-24)
-------------------

- Make source PEP8 compliant. [Leo Singer]

- Update list of GCN notice types. [Leo Singer]

- Fix some lingering Python 3 issues. [Leo Singer]

- Convert ``pygcn-listen`` and ``pygcn-serve`` from scripts into
  console entry points. [Leo Singer]

- Improve unit test coverage. [Leo Singer]

0.1.11 (2018-01-24)
-------------------

- Modernize setuptools usage. [Leo Singer]

- Drop support for Python 2.6, 3.2, and 3.3. [Leo Singer]

0.1.10 (2017-01-05)
-------------------

- Convert command line interface from optparse to argparse. [Leo Singer]

- Add support for Python 3.5 and drop Python 2.6 and 3.1. [Leo Singer]

0.1.9 (2016-07-31)
------------------

- Add AMON_ICECUBE_EHE notice type. [Leo Singer]

0.1.8 (2016-04-27)
------------------

- Add AMON and CALET notice types. [Leo Singer]

0.1.7 (2015-09-13)
------------------

- Cleanly shut down server socket. [Leo Singer]

  This decreases or eliminates the prevalence of ``Address already in use``
  errors when the server is killed (e.g. via a ``KeyboardInterrupt``) and
  immediately restarted.

- Add some log messages. [Leo Singer]

- Update setuptools bootstrap script. [Leo Singer]

0.1.6 (2015-03-26)
------------------

- Update setuptools bootstrap script. [Leo Singer]

0.1.5 (2015-03-26)
------------------

- Add LIGO/Virgo notice types. [Leo Singer]

0.1.4 (2014-05-19)
------------------

- Restore compatibility with Python 2.7.3. [Leo Singer]

  Python 2.7.3's io.BytesIO() cannot accept a bytes object, but not a
  buffer object.

0.1.3 (2014-05-19)
------------------

- Support Python 3.x. [Leo Singer]

- Conditionally import xml.parsers.expat.ExpatError. [Leo Singer]

  Restores compatibility with Python 2.6, which does not have
  xml.etree.cElementTree.ParseError.

- Add unit test for connection closed by server. [Leo Singer]

- Recover from connection closed by peer while reading. [Leo Singer]

- Add nose-based unit tests. Build now uses setuptools. [Leo Singer]

- Fix typos. [Leo Singer]

0.1.2 (2014-02-27)
------------------

- Print error and recover if VOEvent message has no ivorn. [Leo Singer]

- Recover from malformed XML packets. [Leo Singer]

  If a malformed XML packet is received, log its base64-encoded content
  for debugging purposes and then reconnect.

- Add rudimentary server. [Leo Singer]

- Fix spelling error. [Leo Singer]

0.1.1 (2014-02-12)
------------------

- Fix Python 2.6 compatibility. [Leo Singer]

- Prefer lxml.etree over xml.etree. [Leo Singer]

  It's faster.

0.1 (2014-02-07)
----------------

- First release. [Leo Singer]


gcn/notice_types.py
# Copyright (C) 2014-2018  Leo Singer
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
"""
GCN Notice types, from <http://gcn.gsfc.nasa.gov/filtering.html>.
"""

from enum import IntEnum


_notice_types = dict(
    GRB_COORDS=1,
    TEST_COORDS=2,
    IM_ALIVE=3,
    KILL_SOCKET=4,
    MAXBC=11,
    BRAD_COORDS=21,
    GRB_FINAL=22,
    HUNTS_SRC=24,
    ALEXIS_SRC=25,
    XTE_PCA_ALERT=26,
    XTE_PCA_SRC=27,
    XTE_ASM_ALERT=28,
    XTE_ASM_SRC=29,
    COMPTEL_SRC=30,
    IPN_RAW=31,
    IPN_SEG=32,
    SAX_WFC_ALERT=33,
    SAX_WFC_SRC=34,
    SAX_NFI_ALERT=35,
    SAX_NFI_SRC=36,
    XTE_ASM_TRANS=37,
    spare38=38,
    IPN_POS=39,
    HETE_ALERT_SRC=40,
    HETE_UPDATE_SRC=41,
    HETE_FINAL_SRC=42,
    HETE_GNDANA_SRC=43,
    HETE_TEST=44,
    GRB_CNTRPART=45,
    SWIFT_TOO_FOM=46,
    SWIFT_TOO_SC_SLEW=47,
    DOW_TOD=48,
    spare50=50,
    INTEGRAL_POINTDIR=51,
    INTEGRAL_SPIACS=52,
    INTEGRAL_WAKEUP=53,
    INTEGRAL_REFINED=54,
    INTEGRAL_OFFLINE=55,
    INTEGRAL_WEAK=56,
    AAVSO=57,
    MILAGRO_POS=58,
    KONUS_LC=59,
    SWIFT_BAT_GRB_ALERT=60,
    SWIFT_BAT_GRB_POS_ACK=61,
    SWIFT_BAT_GRB_POS_NACK=62,
    SWIFT_BAT_GRB_LC=63,
    SWIFT_BAT_SCALEDMAP=64,
    SWIFT_FOM_OBS=65,
    SWIFT_SC_SLEW=66,
    SWIFT_XRT_POSITION=67,
    SWIFT_XRT_SPECTRUM=68,
    SWIFT_XRT_IMAGE=69,
    SWIFT_XRT_LC=70,
    SWIFT_XRT_CENTROID=71,
    SWIFT_UVOT_DBURST=72,
    SWIFT_UVOT_FCHART=73,
    SWIFT_BAT_GRB_LC_PROC=76,
    SWIFT_XRT_SPECTRUM_PROC=77,
    SWIFT_XRT_IMAGE_PROC=78,
    SWIFT_UVOT_DBURST_PROC=79,
    SWIFT_UVOT_FCHART_PROC=80,
    SWIFT_UVOT_POS=81,
    SWIFT_BAT_GRB_POS_TEST=82,
    SWIFT_POINTDIR=83,
    SWIFT_BAT_TRANS=84,
    SWIFT_XRT_THRESHPIX=85,
    SWIFT_XRT_THRESHPIX_PROC=86,
    SWIFT_XRT_SPER=87,
    SWIFT_XRT_SPER_PROC=88,
    SWIFT_UVOT_POS_NACK=89,
    SWIFT_BAT_ALARM_SHORT=90,
    SWIFT_BAT_ALARM_LONG=91,
    SWIFT_UVOT_EMERGENCY=92,
    SWIFT_XRT_EMERGENCY=93,
    SWIFT_FOM_PPT_ARG_ERR=94,
    SWIFT_FOM_SAFE_POINT=95,
    SWIFT_FOM_SLEW_ABORT=96,
    SWIFT_BAT_QL_POS=97,
    SWIFT_BAT_SUB_THRESHOLD=98,
    SWIFT_BAT_SLEW_POS=99,
    AGILE_GRB_WAKEUP=100,
    AGILE_GRB_GROUND=101,
    AGILE_GRB_REFINED=102,
    SWIFT_ACTUAL_POINTDIR=103,
    AGILE_MCAL_ALERT=105,
    AGILE_POINTDIR=107,
    AGILE_TRANS=108,
    AGILE_GRB_POS_TEST=109,
    FERMI_GBM_ALERT=110,
    FERMI_GBM_FLT_POS=111,
    FERMI_GBM_GND_POS=112,
    FERMI_GBM_LC=113,
    FERMI_GBM_GND_INTERNAL=114,
    FERMI_GBM_FIN_POS=115,
    FERMI_GBM_ALERT_INTERNAL=116,
    FERMI_GBM_FLT_INTERNAL=117,
    FERMI_GBM_TRANS=118,
    FERMI_GBM_POS_TEST=119,
    FERMI_LAT_POS_INI=120,
    FERMI_LAT_POS_UPD=121,
    FERMI_LAT_POS_DIAG=122,
    FERMI_LAT_TRANS=123,
    FERMI_LAT_POS_TEST=124,
    FERMI_LAT_MONITOR=125,
    FERMI_SC_SLEW=126,
    FERMI_LAT_GND=127,
    FERMI_LAT_OFFLINE=128,
    FERMI_POINTDIR=129,
    SIMBADNED=130,
    FERMI_GBM_SUBTHRESH=131,
    SWIFT_BAT_MONITOR=133,
    MAXI_UNKNOWN=134,
    MAXI_KNOWN=135,
    MAXI_TEST=136,
    OGLE=137,
    CBAT=138,
    MOA=139,
    SWIFT_BAT_SUBSUB=140,
    SWIFT_BAT_KNOWN_SRC=141,
    VOE_11_IM_ALIVE=142,
    VOE_20_IM_ALIVE=143,
    FERMI_SC_SLEW_INTERNAL=144,
    COINCIDENCE=145,
    FERMI_GBM_FIN_INTERNAL=146,
    SUZAKU_LC=148,
    SNEWS=149,
    LVC_PRELIMINARY=150,
    LVC_INITIAL=151,
    LVC_UPDATE=152,
    LVC_TEST=153,
    LVC_COUNTERPART=154,
    AMON_ICECUBE_COINC=157,
    AMON_ICECUBE_HESE=158,
    CALET_GBM_FLT_LC=160,
    CALET_GBM_GND_LC=161,
    LVC_EARLY_WARNING=163,
    LVC_RETRACTION=164,
    GWHEN_COINC=168,
    AMON_ICECUBE_EHE=169,
    HAWC_BURST_MONITOR=171,
    AMON_NU_EM_COINC=172,
    ICECUBE_ASTROTRACK_GOLD=173,
    ICECUBE_ASTROTRACK_BRONZE=174,
    SK_SN=175,
    ICECUBE_CASCADE=176)

vars().update(**_notice_types)
NoticeType = IntEnum('NoticeType', _notice_types)
__all__ = ('NoticeType',)
del IntEnum, _notice_types
